### PR TITLE
Add Rust application configuration serializer

### PR DESCRIPTION
**Pull request by ipetr0v**
This change:
- Adds a Rust version of application configuration serializer
- Updates examples to use Rust serializer

Fixes https://github.com/project-oak/oak/issues/740
Ref https://github.com/project-oak/oak/issues/719

# Checklist
- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have checked that these tests are run by [Cloudbuild](/cloudbuild.yaml)

### CODE REVIEW

**Comment by anghelcovici** [comment]
It should be easier to use after #1214. I also have a suggestion for how to change this quite significantly... I was thinking of using a Toml file similar to what @tiziano88 created for the runner as an input for it and drop completely any bash scripts.

**Comment by tiziano88** [comment]
FYI @ipetr0v and I already discussed an app manifest (TOML or whatever format): https://github.com/project-oak/oak/issues/740#issuecomment-651962664

**Comment by tiziano88** [comment]
Oh, it seems @ipetr0v just added it to this PR actually! awesome!

File: examples/abitest/config/config.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name = "abitest"
+
+[modules]
+frontend_module = { path = "examples/target/wasm32-unknown-unknown/release/abitest_0_frontend.wasm" }
+backend_module = { path = "examples/target/wasm32-unknown-unknown/release/abitest_1_backend.wasm" }
+
+[initial_node_configuration]
+wasm_module_name = "app"
```
Comment (tiziano88) [inline examples/abitest/config/config.toml]: I'm confused, shouldn't this value be "frontend_module"?

File: oak/server/rust/oak_config_serializer/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+[package]
```
Comment (tiziano88) [inline oak/server/rust/oak_config_serializer/Cargo.toml:1]: This is really part of the SDK, not of the server, so it should be moved under sdk, or as a separate top-level crate.

File: oak/server/rust/oak_config_serializer/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+use prost::Message;
+use std::{collections::HashMap, fs::read_to_string};
+use structopt::StructOpt;
+
+/// Command line options for the Oak Application Configuration Serializer.
+#[derive(StructOpt, Clone, Debug)]
+#[structopt(about = "Oak Application Configuration Serializer")]
+struct Opt {
+    #[structopt(long, help = "TOML application configuration file")]
+    config_file: String,
```
Comment (tiziano88) [inline oak/server/rust/oak_config_serializer/src/main.rs:43]: Specify somewhere that it is the input file (either in the field name, or in the help)

File: oak/server/rust/oak_config_serializer/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    modules: HashMap<String, Module>,
+    initial_node_configuration: InitialNodeConfig,
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct Module {
+    url: Option<String>,
+    path: Option<String>,
+}
```
Comment (tiziano88) [inline oak/server/rust/oak_config_serializer/src/main.rs:61]: I guess this should really be an enum, but TOML support for enum is not great. Is that why you have two fields?

File: examples/trusted_information_retrieval/config/config.toml [hunk 5]
```diff
@@ -1 +1,9 @@
-database_url = "https://localhost:8888"
+name = "trusted_information_retrieval"
+
+[modules]
+app = { path = "examples/target/wasm32-unknown-unknown/release/trusted_information_retrieval.wasm" }
+database_proxy = { path = "examples/target/wasm32-unknown-unknown/release/database_proxy.wasm" }
+
+[initial_node_configuration]
+wasm_module_name = "app"
+wasm_entrypoint_name = "oak_main"
```
Comment (anghelcovici) [inline examples/trusted_information_retrieval/config/config.toml:9]: This seems like it's repeating everywhere... do we actually need this? Can we just have this as default and only have it when it's different at least?

File: oak/server/rust/oak_config_serializer/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    debug!("parsed opts: {:?}", opt);
+
+    let config_file = read_to_string(opt.config_file)?;
+    let config: Config = toml::from_str(&config_file).context("Couldn't parse TOML config file")?;
+    debug!("parsed config file: {:?}", config);
+
+    // Load Wasm modules.
+    let mut modules = HashMap::new();
+    for (name, module) in config.modules.iter() {
+        modules.insert(name.to_string(), load_module(&module)?);
```
Comment (tiziano88) [inline oak/server/rust/oak_config_serializer/src/main.rs:108]: How about something like


```suggestion
        modules.insert(name.to_string(), load_module(&module).context("could not load module from file {}", module)?);
```

File: oak/server/rust/oak_config_serializer/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            name: config.name,
+            config_type: Some(ConfigType::WasmConfig(WebAssemblyConfiguration {
+                wasm_module_name: config.initial_node_configuration.wasm_module_name,
+                wasm_entrypoint_name: config.initial_node_configuration.wasm_entrypoint_name,
+            })),
+        }),
+    };
+
+    // Serialize application configuration.
+    write_config_to_file(&app_config, &opt.output_file)?;
```
Comment (tiziano88) [inline oak/server/rust/oak_config_serializer/src/main.rs:124]: `.context`

File: examples/trusted_information_retrieval/config/config.toml [hunk 8]
```diff
@@ -1 +1,9 @@
-database_url = "https://localhost:8888"
+name = "trusted_information_retrieval"
+
+[modules]
+app = { path = "examples/target/wasm32-unknown-unknown/release/trusted_information_retrieval.wasm" }
```
Comment (anghelcovici) [inline examples/trusted_information_retrieval/config/config.toml:4]: I find it inelegant that this is referencing wasm modules via full target path. cargo has an experimental option to select an output folder for the binary. In #1214 I have out the config files in a "bin" folder. Maybe we should try the same with the wasm files?

File: examples/trusted_information_retrieval/config/config.toml [hunk 9]
```diff
@@ -1 +1,9 @@
-database_url = "https://localhost:8888"
+name = "trusted_information_retrieval"
```
Comment (anghelcovici) [inline examples/trusted_information_retrieval/config/config.toml:1]: What is the name for?

File: oak_utils/src/io.rs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+/// Read the contents of a file.
+pub fn read_file(filename: &str) -> anyhow::Result<Vec<u8>> {
+    let mut file =
+        File::open(filename).with_context(|| format!("Failed to open file <{}>", filename))?;
+    let mut buffer = Vec::new();
+    file.read_to_end(&mut buffer)
+        .with_context(|| format!("Failed to read file <{}>", filename))?;
+    Ok(buffer)
```
Comment (tiziano88) [inline oak_utils/src/io.rs:30]: I think https://doc.rust-lang.org/std/fs/fn.read.html does this already

File: oak_utils/src/io.rs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+/// Write byte vector into a file (rewrite if file exists).
+pub fn write_file(data: &Vec<u8>, filename: &str) -> anyhow::Result<()> {
+    let mut file =
+        File::create(filename).with_context(|| format!("Failed to create file <{}>", filename))?;
+    file.write_all(&data)
+        .with_context(|| format!("Failed to write file <{}>", filename))?;
+    Ok(())
```
Comment (tiziano88) [inline oak_utils/src/io.rs:39]: https://doc.rust-lang.org/std/fs/fn.write.html

File: scripts/build_example [hunk 12]
```diff
@@ -53,6 +53,7 @@ if [[ -z "${EXAMPLE+z}" ]]; then
   exit 1
 fi
 
+mkdir -p "out/examples/${EXAMPLE}"
```
Comment (tiziano88) [inline scripts/build_example:56]:
```suggestion
mkdir --parents "out/examples/${EXAMPLE}"
```

File: examples/abitest/config/config.toml (hunk 1 above)
Comment (ipetr0v) [reply examples/abitest/config/config.toml]: Done

File: examples/trusted_information_retrieval/config/config.toml (hunk 9 above)
Comment (ipetr0v) [reply examples/trusted_information_retrieval/config/config.toml:1]: To assign here:
https://github.com/project-oak/oak/blob/c15f2858377f6944f0c0fce4d31b2481e70d8518/oak/proto/application.proto#L54-L58

File: examples/trusted_information_retrieval/config/config.toml (hunk 5 above)
Comment (tiziano88) [reply examples/trusted_information_retrieval/config/config.toml:9]: I suggest to stick with this for now, we can factor it out if it gets annoying in the future, but it's another piece of documentation that developers need to find (what is the default module name? what should the function be called? etc.)

File: examples/trusted_information_retrieval/config/config.toml (hunk 8 above)
Comment (tiziano88) [reply examples/trusted_information_retrieval/config/config.toml:4]: +1 for some sort of `out` / `bin` folder

**Comment by anghelcovici** [comment]
@ipetr0v  - In case it wasn't clear, just make sure you wait for #1214, as you should probably build on that.

File: oak_utils/src/io.rs (hunk 10 above)
Comment (ipetr0v) [reply oak_utils/src/io.rs:30]: Done

File: oak_utils/src/io.rs (hunk 11 above)
Comment (ipetr0v) [reply oak_utils/src/io.rs:39]: Done

File: examples/trusted_information_retrieval/config/config.toml (hunk 8 above)
Comment (ipetr0v) [reply examples/trusted_information_retrieval/config/config.toml:4]: Will rewrite to use `bin` once #1214 is merged.

File: oak/server/rust/oak_config_serializer/src/main.rs (hunk 4 above)
Comment (ipetr0v) [reply oak/server/rust/oak_config_serializer/src/main.rs]: Done
Using `enum`

File: sdk/rust/oak_config_serializer/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+impl Default for InitialNodeConfig {
+    fn default() -> Self {
+        Self {
+            wasm_module_name: "app".to_string(),
+            wasm_entrypoint_name: "oak_main".to_string(),
+        }
+    }
+}
```
Comment (tiziano88) [inline sdk/rust/oak_config_serializer/src/main.rs:79]: Does this mean that it can be omitted from the TOML files?

Comment (ipetr0v) [reply sdk/rust/oak_config_serializer/src/main.rs:79]: Done

File: examples/trusted_information_retrieval/config/config.toml (hunk 8 above)
Comment (ipetr0v) [reply examples/trusted_information_retrieval/config/config.toml:4]: Done

File: scripts/build_example (hunk 12 above)
Comment (daviddrysdale) [reply scripts/build_example]: By the way, expanding every argument isn't always helpful &ndash; this fails on macOS.

**Pull request merged**